### ISSUE TITLE
Trying to interject authorization, but context is always nil

### ISSUE THREAD

**Issue by sirvine** [JSONAPI-Resources/jsonapi-resources#458]
Following the README, I'm trying to interject an authorization step across all resources. I've got a base resource that looks like this:

```
    module Api
      module V1
        class BaseResource < JSONAPI::Resource
          def records_for(relation_name, options={})
            context = options[:context]
            records = model.public_send(relation_name)

            # just logging options hash for the time being
            Rails.logger.debug options

            records
          end
        end
      end
```

All of my other resources inherit from this `BaseResource` class. I'm just passing the records through and logging the `options` hash for now. `options` is always `{:context => nil}`.

I also have a base controller called `API::V1::ApplicationController`, which inherits from `JSONAPI::ResourceController`, and which includes the following method:

```
def context
  {current_user: current_user}
end
```

All of my other API controllers inherit from `API::V1::ApplicationController`. I can confirm that `current_user` is set properly when the `context` method is called, but it never makes its way to the options hash in `BaseResource`.

**Comment by sirvine** [comment]
After posting this issue, I notice that the context comes through only when `get_related_resources` is called from the client. I'm still confused about how to push the current_user into a resource for all requests that hit it.

**Comment by lgebhardt** [comment]
@sirvine I'm not sure exactly what is wrong, but I have some ideas. I've created a PR #459 that cleans up the way resource instances are initialized with the context. This also removes the options hash from `records_for` since it really isn't needed.

I'm wondering if you have a resource being created, say in an overridden `find_by_key` method, that doesn't provide the context in the initializer. I'm hoping this fixes your problem by exposing that sort of scenario, but I'm not sure.

If you could test against this PR that would be great. When you do you will need to modify your code to use @context:

``` ruby

    module Api
      module V1
        class BaseResource < JSONAPI::Resource
          def records_for(relation_name)
            records = model.public_send(relation_name)

            # just logging @context for the time being
            Rails.logger.debug @context

            records
          end
        end
      end
```

**Comment by sirvine** [comment]
Your PR led me to a solution. I had two convenience methods for rendering serialized results in my `API::V1::ApplicationController`:

```
def jsonapi_collection(collection, resource_klass, options={})
  jsonapi_resources = []
  collection.each do |item|
    jsonapi_resources.push resource_klass.new(item)
  end
  JSONAPI::ResourceSerializer.new(resource_klass, options).serialize_to_hash(jsonapi_resources)
end

def jsonapi_item(item, resource_klass, options={})
  JSONAPI::ResourceSerializer.new(resource_klass, options).serialize_to_hash(resource_klass.new(item))
end
```

I was calling these in the controllers, as follows:

```
 render json: jsonapi_collection(@users, UserResource, {include: ['clients']})
```

Under your PR, I had to modify these convenience methods to pass the options hash, or else I got an error for a missing argument:

```
def jsonapi_collection(collection, resource_klass, options={})
  jsonapi_resources = []
  collection.each do |item|
    jsonapi_resources.push resource_klass.new(item, options)
  end
  JSONAPI::ResourceSerializer.new(resource_klass, options).serialize_to_hash(jsonapi_resources)
end

def jsonapi_item(item, resource_klass, options={})
  JSONAPI::ResourceSerializer.new(resource_klass, options).serialize_to_hash(resource_klass.new(item, options))
end
```

(Note the second `options` argument.)

That got me thinking that these convenience methods might be interfering with the normal operation of the `context` method. Indeed, modifying them as follows fixed my issue:

```
def jsonapi_collection(collection, resource_klass, options={})
  jsonapi_resources = []
  collection.each do |item|
    jsonapi_resources.push resource_klass.new(item, options)
  end
  options[:current_user] = current_user
  JSONAPI::ResourceSerializer.new(resource_klass, options).serialize_to_hash(jsonapi_resources)
end

def jsonapi_item(item, resource_klass, options={})
  options[:current_user] = current_user
  JSONAPI::ResourceSerializer.new(resource_klass, options).serialize_to_hash(resource_klass.new(item, options))
end
```

(Note the explicit insertion of `current_user` into the `options` hash.)

It's super late here in Japan, so I'm probably too foggy-headed to understand what happened here, but it's working now. Let me know how you want to handle the PR, and I'll adjust accordingly.

Thanks!

**Comment by lgebhardt** [comment]
Glad that you were able to get it working. We'll merge the PR soon and I'll get a 0.6.0 release out after that.

As to what happened, resources store the context as an instance variable that was set in the initializer. If you only use resources through the "normal" channels you won't need to think about this. But as soon as you create a resource yourself you needed to make sure you passed the context in so it could store it in the instance. If you didn't it defaulted to nil and this was used when fetching related resources through the `resource_for` method. The PR makes the context a required parameter, which as you found out, you weren't providing (because the code allowed it and then just failed silently).

**Comment by beechnut** [comment]
Hi @sirvine! I just had the same issue, and couldn't figure out what was going on. I'm glad you figured out a way around it, but we stumbled on another answer that I hope can be of help.

In the [`ApplicationController`](https://github.com/MAPC/know-place-api/blob/develop/app/controllers/application_controller.rb#L9), we had

``` ruby
before_action :authenticate_user_from_token
```

We found what you did: the controllers had the correct `current_user`, but it wasn't getting passed into the context. I think this was because `#context` was being called before `current_user`

In a moment of desperation, we replaced `before_action` with `prepend_before_action`:

``` ruby
prepend_before_action :authenticate_user_from_token
```

I won't claim to understand why this works, but it did for us. It's probably not a permanent solution to this issue, but maybe it can provide some insight for the next release.

**Comment by lgebhardt** [comment]
@beechnut By using `prepend_before_action` you are pushing your filter to the front of the line. Because JR also uses filters to setup the request it's possible one will be called before the other, unless you use `prepend_before_action`. I will look into an alternate way of setting up the request to not interfere with the filter chain.

**Comment by sirvine** [comment]
Indeed, I had `before_filter :authenticate_user_from_token!` on my resource-specific controllers, as per devise documentation.

**Comment by brancusi** [comment]
Hi all, having this issue I'm assuming. I am using devise and ember simple auth. Context does not have `current_user`. Have tried both `prepend_before_action` and `before_filter`.

Any ideas?

**Comment by beechnut** [comment]
@brancusi: It looks like your issue was solved in #580. Is that correct?

**Comment by brancusi** [comment]
@beechnut yes, thank you!
